Merge in 64 bit types branch to master

Cekis opened the pull request:
Ready to be reviewed.  This work is mostly for a future 64 bit transition in the future, but all work has been made compatible with our 32 bit builds.  Has been tested by QA as well (@langelusse).

Cekis committed:
Compiling 100% with Types change approach - still does not build.

Cekis committed:
Added Crypto++ v8.6 - does not yet build as a whole

Cekis committed:
Progress on updating Crypto++

Cekis committed:
Adjusting CMake files

Cekis committed:
Compiling 100% - Execution still broken - DataTable tool may need adjustment

Cekis committed:
Datatables now compiling correctly

Cekis committed:
Fixed issues loading string tables.

Cekis committed:
Fixed issue with profanity filter

Cekis committed:
Fixed timeout issue with server startup.

Cekis committed:
Removed native methods that no longer exist.

Cekis committed:
Fixed some incorrect sizing on network messages.

Cekis committed:
Reworked Archive package for standard data types and fixed issue with Server package overrun

Cekis committed:
Fixed issue with PlayerQuestData parsing - server is RUNNING

Cekis committed:
Updated readme and added recognition to Apathy

Mega thanks to Apathy for his help in making the 64 bit version happen.  Without his knowledge and super-special ability to act as a mega sounding board.  Thanks for letting me ramble and lose my mind!

Cekis committed:
Update README.md

Cekis committed:
Adjusted a few more data types to fix communication issues.

Cekis committed:
Merge branch '64-bit-types' of https://github.com/SWG-Source/src into 64-bit-types

Cekis committed:
Fixed size issue with ByteStream and removed some debug output

Cekis committed:
Back port to 32 bit to take advantage of 64 bit adjustments

Cekis committed:
Updating badly chosen replacement data type

Cekis committed:
Removed old crypto lib

Cekis committed:
Added adjustments that allow compiling with CLang 9.0.1 or GCC 10

Cekis committed:
Fixed compiler warnings for mixing Strings with char *

Cekis committed:
Added Oracle 21 paths and fixed issue with naming

Cekis committed:
Fixed issues with CMake configuration

Cekis committed:
Fixed issue for newer compilers and adjusted Oracle and chat for compilation

Cekis committed:
Updated CMake build flags

Cekis committed:
Adjusted DbBindableLong to be 64 bits if needed and adjusted timing for 64 bits

Cekis committed:
Refactored DbBindableLong to be more appropriately named DbBindableInt32

Cekis committed:
Updating CMake build files to be more reliable and readable.

Cekis committed:
Added comments and removed setting defaults again for CMake

Cekis committed:
Addressed several compiler warnings from GCC

Cekis committed:
Merged in master branch

AlecH92 reviewed (approved):
Looks good to me. I skipped over the new Crypto library files.

Cekis commented:
@AlecH92 Cool... yes, most of the line changes are from adding the updated Crypto++ Lib.  Thanks for looking it over!

AconiteX commented on engine/server/library/serverGame/src/shared/ai/HateList.cpp:
```diff
@@ -535,8 +535,8 @@ void HateList::addServerNpAutoDeltaVariables(Archive::AutoDeltaByteStream & stre
-	time_t const timeSinceActivitiy = (currentTime - m_lastUpdateTime.get());
+	uint32_t const currentTime = Os::getRealSystemTime();
+	uint32_t const timeSinceActivitiy = (currentTime - m_lastUpdateTime.get());
 
```
so here we're replacing time_t in which case we still want to be using uint64_t but `Os::getRealSystemTime()` is still a time_t as it wasn't adjusted in this changeset

AconiteX commented on engine/server/library/serverGame/src/shared/command/CommandCppFuncs.cpp:
```diff
@@ -1800,11 +1800,11 @@ static void commandFuncSpatialChatInternal(Command const &, NetworkId const &act
 					{
-						GenericValueTypeMessage<int> csl("ChatSpamLimited", (chatSpamTimeEndInterval - timeNow));
+						GenericValueTypeMessage<uint32_t> csl("ChatSpamLimited", (chatSpamTimeEndInterval - timeNow));
 						obj->getClient()->send(csl, true);
```
more time stuff

AconiteX commented on engine/server/library/serverGame/src/shared/controller/AiCreatureController.cpp:
```diff
@@ -2960,7 +2960,7 @@ PersistentCrcString const & AICreatureController::getCombatAction()
 //-----------------------------------------------------------------------
-time_t AICreatureController::getKnockDownRecoveryTime() const
+uint32_t AICreatureController::getKnockDownRecoveryTime() const
 {
```
time

AconiteX commented on engine/server/library/serverGame/src/shared/controller/AiCreatureController.h:
```diff
@@ -186,7 +186,7 @@ class AICreatureController : public CreatureController
 	PersistentCrcString const & getCombatAction();
-	time_t getKnockDownRecoveryTime() const;
+	uint32_t getKnockDownRecoveryTime() const;
 
```
time

AconiteX commented on engine/server/library/serverGame/src/shared/core/CharacterMatchManager.cpp:
```diff
@@ -639,14 +639,14 @@ void CharacterMatchManager::requestMatch(NetworkId const &networkId, MatchMaking
 
-	const unsigned long endTimeMs = Clock::timeMs();
+	const uint32_t endTimeMs = Clock::timeMs();
 	if (endTimeMs >= startTimeMs)
```
time

AconiteX commented on engine/server/library/serverGame/src/shared/core/GameServer.cpp:
```diff
@@ -4059,17 +4060,17 @@ void GameServer::run(void)
-	unsigned long lastFrameProcessStartTime = startTime;
+	uint32_t startTime = Clock::timeMs();
+	uint32_t lastFrameProcessStartTime = startTime;
 
```
time

AconiteX commented on engine/server/library/serverNetworkMessages/src/shared/TaskConnectionIdMessage.h:
```diff
@@ -34,15 +34,15 @@ class TaskConnectionIdMessage : public GameNetworkMessage
 	const std::string &  getClusterName  () const;
-	const long           getCurrentEpochTime() const;
+	const int32_t        getCurrentEpochTime() const;
 private:
```
time

AconiteX commented on engine/server/library/serverPathfinding/src/shared/CityPathGraph.cpp:
```diff
@@ -709,32 +709,32 @@ void CityPathGraph::setLinkDistance ( float dist )
+//
+//	return code;
+//}
 
```
we commenting this out bc? 
idk if i even what to know wtf this monster is

AconiteX commented on engine/server/library/serverScript/src/shared/ScriptMethodsCollection.cpp:
```diff
@@ -90,7 +90,7 @@ bool ScriptMethodsCollectionNamespace::install()
 		JF("getCollectionSlotName", "(I)Ljava/lang/String;", getCollectionSlotName),
-		JF("getCollectionSlotMaxValue", "(Ljava/lang/String;)J", getCollectionSlotMaxValue),
+		//JF("getCollectionSlotMaxValue", "(Ljava/lang/String;)J", getCollectionSlotMaxValue),
 		JF("getAllCollectionSlotsInCollection", "(Ljava/lang/String;)[Ljava/lang/String;", getAllCollectionSlotsInCollection),
```
this is commenting but just reiterating it's used in 3.1 so will need to be fixed

AconiteX commented on engine/server/library/serverScript/src/shared/ScriptMethodsCombat.cpp:
```diff
@@ -1680,7 +1680,7 @@ jboolean JNICALL ScriptMethodsCombatNamespace::doCombatResults(JNIEnv *env, jobj
 	{
-		WARNING(true, ("doCombatResults(): caller passed in unsupported combat action name [%s]!  Continuing to send combat action.  Call stack follows in stdout.", animationString.c_str()));
+		DEBUG_WARNING(true, ("doCombatResults(): caller passed in unsupported combat action name [%s]!  Continuing to send combat action.  Call stack follows in stdout.", animationString.c_str()));
 		// @todo make this go to the log rather than stdout; otherwise, we're going to have to do a lot more work to see this info.
```
why are we dropping this to DEBUG? is this getting thrown a lot?

AconiteX commented on engine/server/library/serverScript/src/shared/ScriptMethodsPermissions.cpp:
```diff
@@ -66,7 +66,7 @@ const JNINativeMethod NATIVES[] = {
 	JF("_expelFromBuilding", "(J)V", expelFromBuilding),
-	JF("_sendDirtyCellPermissionsUpdateToClient", "(JJZ)V", sendDirtyCellPermissionsUpdateToClient),
+	//JF("_sendDirtyCellPermissionsUpdateToClient", "(JJZ)V", sendDirtyCellPermissionsUpdateToClient),
 };
```
3.1

AconiteX commented on engine/server/library/serverScript/src/shared/ScriptMethodsSystem.cpp:
```diff
@@ -50,7 +50,7 @@ const JNINativeMethod NATIVES[] = {
 	JF("_playCutScene", "(JLjava/lang/String;)V", playCutScene),
-	JF("_triggerServerWarning", "(Ljava/lang/String;)V", triggerServerWarning),
+	// JF("_triggerServerWarning", "(Ljava/lang/String;)V", triggerServerWarning),
 };
```
3.1

AconiteX commented on engine/server/library/serverUtility/src/shared/ChatLogManager.cpp:
```diff
@@ -73,14 +73,14 @@ namespace ChatLogManagerNamespace
 		Unicode::String m_channel;
-		time_t m_time;
+		uint32_t m_time;
 
```
time everywhere

AconiteX commented on engine/shared/library/sharedFoundation/src/shared/NetworkId.h:
```diff
@@ -68,8 +68,8 @@ class NetworkId
-	explicit NetworkId(int value); // catch people who incorrectly try to construct one of these from an int
+//  private:
+//	explicit NetworkId(int value); // catch people who incorrectly try to construct one of these from an int
 	// also makes constructing one of these from any other incorrect numeric type ambiguous
```
are we replacing this with some sort of verification that will throw an error if we pass an int to a NetworkId constructor otherwise we're just eliminating the type safety?

AconiteX commented on engine/shared/library/sharedNetworkMessages/src/shared/common/GameNetworkMessage.cpp:
```diff
@@ -122,7 +122,7 @@ std::string const &GameNetworkMessage::getCmdName() const
 
-std::string const &GameNetworkMessage::getCmdName(unsigned long cmdCrc) // static
+std::string const &GameNetworkMessage::getCmdName(uint32_t cmdCrc) // static
 {
```
didn't we talk about crc being weirdly sometimes signed and sometimes unsigned

Cekis replied:
Yes, however, this particular method used the unsigned version.  Just keeping it as it is (unsigned) such that we don't break anything.

AconiteX commented on engine/shared/library/sharedObject/src/shared/object/AlterScheduler.cpp:
```diff
@@ -111,10 +111,10 @@ namespace AlterSchedulerNamespace
 
-	uint32 const cs_freeFillPattern         = 0xEFEFEFEF; // this should match MemoryManager's free fill pattern.
+	uint64_t const cs_freeFillPattern         = 0xEFEFEFEFEFEFEFEF; // this should match MemoryManager's free fill pattern.
 	
```
The ref to MemoryManager is probably(?) obsolete given the deprecation of MemoryManager

AconiteX commented on engine/shared/library/sharedObject/src/shared/object/AlterScheduler.cpp:
```diff
@@ -111,10 +111,10 @@ namespace AlterSchedulerNamespace
 	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 
-	float const  cs_schedulerTicksPerSecond = 1000.0f;  // # schedule ticks per second.  Frame rates will not be able to exceed this.
+	float const  cs_schedulerTicksPerSecond = 2000.0f;  // # schedule ticks per second.  Frame rates will not be able to exceed this.
```
We should probably config value this

Cekis replied:
Agreed.

AconiteX commented on engine/shared/library/sharedRandom/src/shared/FastRandomGenerator.h:
```diff
@@ -16,18 +16,18 @@ class FastRandomGenerator
 {
 public:
```
this whole thing can probably be replaced with a std native lib since its not 1999 STLPort anymore

AconiteX commented on external/3rd/library/soePlatform/VChatAPI/utils2.0/utils/Base/serializeClasses.h:
```diff
@@ -389,57 +389,45 @@ namespace soe
+		stream[BYTE6] = (data>>40)&0xff;
+		stream[BYTE7] = (data>>48)&0xff;
+		stream[BYTE8] = (data>>56)&0xff;
 		return sizeof(int64);
```
VChatAPI is deprecated, isn't it? can we just delete whatever cancer this is

Cekis replied:
Deprecated?  What's being used in replacement?  Discord?  Would be an effort to remove it properly.

AconiteX replied:
Nothing is being used as its replacement we don't have the Vivox libs so there's no support for Voice Chat and even to re-add it with something like Discord we wouldn't use the VChatAPI wrapper afaik it'd just be through ChatServer handling.

Cekis commented on engine/server/application/LoginServer/src/shared/LoginServer.cpp:
```diff
@@ -480,7 +480,9 @@ void LoginServer::receiveMessage(const MessageDispatch::Emitter &source, const M
+                if(!found) {
+                    DEBUG_REPORT_LOG(true, ("Tried to remove a connection server that wasn't in our list.\n"));
+                }
                 m_clusterStatusChanged = true;
```
Because the DEBUG_REPORT_LOG() doesn't "use" the `found` var so the compiler thinks it's a wasted value.  Doing this "uses" the `found` var and therefore shuts up the compiler.

Cekis commented on engine/server/application/LoginServer/src/shared/SessionApiClient.cpp:
```diff
@@ -120,7 +120,7 @@ void SessionApiClient::OnSessionValidate(const apiTrackingNumber trackingNumber,
 	{
 		// attempt to validate a CS Tool associated with this request.
-		CSToolConnection::validateCSTool( ( uint32 )userData, result, session );
+		CSToolConnection::validateCSTool( ( uint64 )userData, result, session );
```
This likely needs updating a bit deeper.  It's being cast as a uint64 because the calling method has userData as a `void *` type.  Because of 64 bits, a `void *` becomes a 64 bit value.  The calling method needs to be updated to not accept a void if at all possible and all callers of the calling method should also be updated.  Great catch!

Cekis commented on engine/server/application/LoginServer/src/shared/TaskCreateCharacter.h:
```diff
@@ -33,12 +33,12 @@ class TaskCreateCharacter : public DB::TaskRequest
+		DB::BindableInt32         cluster_id; //lint !e1925 // public data member
+		DB::BindableInt32         station_id; //lint !e1925 // public data member
 		DB::BindableString<127>  character_name; //lint !e1925 // public data member
 		DB::BindableNetworkId    character_id; //lint !e1925 // public data member
```
It honestly could be - however, after looking at this closer, BindableNetworkId isn't a straight mirror of BindableInt64.  There are manipulations done to it after receiving the value.

Cekis committed:
Updated method as userData is actually a 32 bit integer

Cekis commented on engine/server/application/LoginServer/src/shared/TaskGetValidationData.cpp:
```diff
@@ -26,10 +26,10 @@ namespace TaskGetValidationDataNamespace
+		DB::BindableInt32 station_id; //lint !e1925 // public data member
+		DB::BindableInt32 cluster_id; //lint !e1925 // public data member
+		DB::BindableInt32 character_type_id; //lint !e1925 // public data member
+		DB::BindableInt32 num_open_slots; //lint !e1925 // public data member
```
BindableInt32 handles anything 32 bits and lower.  There's no BindableInt16 or BindableInt8 classes

Cekis commented on engine/server/application/TaskManager/src/shared/ManagerConnection.cpp:
```diff
@@ -103,8 +103,8 @@ void ManagerConnection::onReceive(const Archive::ByteStream & message)
-			GenericValueTypeMessage<std::pair<std::string, long > > msg(r);
+			int32_t const currentTime = static_cast<int32_t>(::time(nullptr));
+			GenericValueTypeMessage<std::pair<std::string, int32_t > > msg(r);
 
```
The client is expecting a value of 32 bits.  Unfortunately, we have a year 2038 problem until we get 64 bit all buckled down on the server and the client transitions to a 64 bit client.  Once that happens, we can adjust the time values to a time_t size.

Cekis commented on external/ours/library/localization/src/shared/LocalizedStringTable.h:
```diff
@@ -46,7 +46,7 @@ class LocalizedStringTable
 
 	// TODO: make this typedef platform dependent
-	typedef long                  magic_type;
+	typedef uint32_t              magic_type;
```
should be signed, not unsigned

Cekis commented on external/ours/library/localizationArchive/src/shared/StringIdArchive.cpp:
```diff
@@ -23,7 +23,7 @@ namespace Archive
 	void get(ReadIterator & source, StringId & target)
 	{
-		unsigned long index = 0;
+		int32_t index = 0;
```
Should be unsigned, not signed.

Cekis commented on game/server/application/SwgDatabaseServer/src/shared/cleanup/TaskObjectTemplateListUpdater.cpp:
```diff
@@ -32,7 +32,7 @@ bool TaskObjectTemplateListUpdater::process(DB::Session *session)
 		while(1)
 		{
-			int i_id;
+			uint32 i_id;
```
Should this be signed instead?

Cekis reviewed (commented):
These changes need to be made or in some cases just verified.

Cekis commented on engine/server/library/serverGame/src/shared/object/ResourcePoolObject.cpp:
```diff
@@ -54,7 +54,7 @@ ResourcePoolObject::~ResourcePoolObject()
 float ResourcePoolObject::harvest(float installedEfficiency, uint32 lastHarvestTime) const
 {
-	int elapsedTime=static_cast<int>(std::min(ServerClock::getInstance().getGameTimeSeconds(),m_depletedTimestamp) - lastHarvestTime);
+	int elapsedTime=static_cast<int>(std::min(ServerClock::getInstance().getGameTimeSeconds(),static_cast<unsigned long>(m_depletedTimestamp)) - lastHarvestTime);
```
This is fine - std::min needs it to be a long (because `getGameTimeSeconds()` is a long) so it's a quick cast only.

Cekis commented on engine/server/library/serverNetworkMessages/src/shared/gameConnectionServer/GameConnectionServerMessages.cpp:
```diff
@@ -86,7 +86,7 @@ TransferControlMessage::~TransferControlMessage()
 ControlAssumed::ControlAssumed(NetworkId oid, const std::string & newSceneName, const bool skipLoadScreen,
 							   const Vector &startPos, const float startYaw, const std::string & templateName,
-							   const int64 t) : 
+							   const int32_t t) :
```
backing this out and making it 64 bit again considering the client is expecting it.

Cekis commented on engine/shared/library/sharedGame/src/shared/core/TextManager.cpp:
```diff
@@ -297,8 +297,8 @@ Unicode::String TextManager::filterText(Unicode::String const &text)
 	Unicode::String::const_iterator iterText = text.begin();
-	int startIndex = Unicode::String::npos;
-	int currentIndex = 0;
+	size_t startIndex = Unicode::String::npos;
```
Unicode::String::npos returns a size_t value.  This is scoped only to this function and is only used in 64 bit context.

Cekis commented on engine/shared/library/sharedDatabaseInterface/src_oci/DbBindableVarray.cpp:
```diff
@@ -120,7 +120,7 @@ OCIType* BindableVarray::getTDO()
 // ----------------------------------------------------------------------
 
-bool BindableVarrayNumber::push_back(int value)
+bool BindableVarrayNumber::push_back(int16_t value)
```
just an ordering thing in the commit.  This isn't truly going from 32 to 16.  There are methods that handle both 32 and 16.

Cekis commented on engine/shared/library/sharedFoundation/src/shared/Misc.h:
```diff
@@ -230,7 +230,7 @@ inline void *memmove(void *destination, const void *source, int length)
 	DEBUG_FATAL(!destination, ("nullptr destination arg"));
 	DEBUG_FATAL(!source, ("nullptr source arg"));
-	return memmove(destination, source, static_cast<uint>(length));
+	return memmove(destination, source, static_cast<size_t>(length));
```
Also, seems odd that we would need to cast here.  Removing the cast.

Cekis commented on engine/shared/library/sharedGame/src/shared/core/TextManager.cpp:
```diff
@@ -120,7 +120,7 @@ void TextManagerNamespace::getAppropriateWord(Unicode::String &text)
 
 			Unicode::String lowerText(Unicode::toLower(text));
-			unsigned int findStartPosition = 0;
+			size_t findStartPosition = 0;
```
Should be ok - just used as a very large potential index of a bad word in a given phrase.

Cekis replied:
actually... going back to uint32_t

Cekis committed:
Fixed issues found in my own review.

Cekis committed:
Adjustments to issues when compiling in 64 bits

Cekis committed:
Backed out bad commit and added cleaner logging to mIFF.

Cekis committed:
Removed extra line break on mIFF compiler

Cekis committed:
Interim commit checkpoint - starts up but failure to read data after quest progresses

Cekis committed:
Merge branch 'master' into 64-bit-types

Cekis committed:
Added static casts to BindableLongs when storing data

Cekis committed:
Organized CMake file a bit better

Cekis committed:
Removed problematic compiler flags

Cekis committed:
Fixed bugs related to database inconsistencies between 32 bit and 64 bit.

swgsais committed:
Support Oracle 19 x64 Instant Client discovery

swgsais committed:
Use fixed-width counts in the NetworkId and Unicode packed maps

These three AutoDeltaPackedMap specializations wrote their command count as
int32 but wrote the baseline count, and read both counts, as size_t. Under
ILP32 size_t was 4 bytes so pack and unpack agreed. Under LP64 size_t binds
to the 8-byte uint64 Archive overload, so pack writes 4+8 bytes while unpack
reads 8+8 and leaves the stream misaligned by 4 bytes for every subsequent
key and value.

Switch both counts to int32_t, matching the generic internal_pack/
internal_unpack path and the already-converted PlayerQuestData specialization.
This also keeps the serialized form identical to the 32-bit server, so data
written by either architecture stays readable by the other.

swgsais committed:
Store call stack addresses at pointer width

DebugHelp::getCallStack() handed the caller's uint32 buffer straight to
backtrace(), which writes void* entries. Under LP64 that wrote 8 bytes per
frame into a 4-byte-per-frame array, overrunning every buffer by 2x:

  Fatal.cpp        uint32[68] = 272 bytes, up to 544 written (stack)
  CallStack        uint32[32] = 128 bytes, 208 written (heap)
  CallStackCollector uint32[26] = 104 bytes, 208 written (heap)

Fatal.cpp's formatMessage() backs every WARNING and FATAL, so any warning
that captured a stack corrupted memory. Observed as SIGSEGV inside cfree()
with si_addr 0x8 after an unrelated warning fired.

Capture into a native void* array and widen into a uint64 buffer, which is
correct under both ILP32 and LP64, and widen the storage and printing to
match. lookupAddress() already took uint64 on linux; win32 now agrees and no
longer truncates StackWalk's DWORD64 address to 32 bits.

Also fixes CallStackCollector copying only one element instead of the whole
call stack, and CRCing only half the buffer.

swgsais committed:
Back QueryPerformanceCounter with a monotonic clock

The linux shim sampled gettimeofday(), which reports CLOCK_REALTIME and can
be slewed or stepped backwards by time synchronisation. Clock::update()
subtracts successive samples to derive frame time, so a backwards step
produced a negative elapsed time, which AlterScheduler then reported as
"no forward time movement" once per affected frame.

Use clock_gettime(CLOCK_MONOTONIC) and keep microsecond units so the value
still matches QueryPerformanceFrequency(). Falls back to gettimeofday() if
the call fails. All getFrameStartTimeMs() consumers compare values
relatively, so nothing depends on the previous epoch-based origin.

swgsais committed:
Stop dereferencing a 32-bit frame offset in the SIGSEGV handler

JavaLibrary::fatalHandler() read the faulting address from a hardcoded 0x44
byte offset into its own frame. That constant describes the 32-bit x86 frame
layout; on x86-64 it is misaligned and points at unrelated stack memory, and
dereferencing it from inside a SIGSEGV handler risks a recursive fault that
loses the crash report entirely. The original code carried a TODO noting the
offset was never verified.

Restrict that probe to __i386__ and rely on the portable
__builtin_return_address() probe already present below it. Zero initialise
the lib and file buffers, which the strstr() checks read whenever a lookup
fails.

AconiteX committed:
Merge pull request #34 from Galaxies-Reborn/x64-dx9-no-docker

Fix four LP64 defects and add Oracle 19 client discovery
